### DISCUSSION TITLE
Examples/guide on how to implement basic custom cost model?

### DISCUSSION THREAD

**Discussion by nvcnvn** [valhalla/valhalla#4180] [Q&A] [answered]
https://valhalla.github.io/valhalla/sif/dynamic-costing/

This surely a newbie question that not very familiar with C++ ecosystem.

Basically I hope we can have step by step examples on how I can create my own class that implement "EdgeCost" function and register that class to Sif and hopefully the rest API will recognize my custom cost method name.

**Comment by kevinkreiser** [comment]
the best guide is a previous pr doing this: https://github.com/valhalla/valhalla/pull/1756/files

it may look like a lot and thats because the idea of a new costing model is a fairly important distinction. im not sure what you are looking to do exactly but it might be worth you telling us a bit more in case it might make sense to just take an existing model and add some new functionality to it. at any rate, the choice is yours!

**Reply by nvcnvn** [reply]
Hi @kevinkreiser thanks for the quick reply.
Do we have any methods to adjust the cost via configuration file for example?

In the long term, I want to implement a costing model based on historical traffic history, (like something mentioned in this article https://valhalla.github.io/valhalla/thor/simple_traffic/), and maybe our Data Scientist team have some magic model that I can make a rpc call to adjust the cost.

But in short term, I have a small project to adjust the cost of some particular street based on some known rules, for example I know some roads will be very busy depend on time/day/direction...

**Reply by kevinkreiser** [reply, marked as answer]
we already support historical traffic and if you know which roads you can give them worse free flow and constrained speeds via valhalla_add_predictive_traffic the documentation is non existant but the unit tests basically show how to do it.

**Reply by nvcnvn** [reply]
Thanks you, that amazing to know!

**Reply by ChouBokYann** [reply]
Hi @kevinkreiser, with regards to this. My team is also hoping to update the costing model with historical traffic history but we came across a road bump when updating the same segment but in different directions. From what we have seen so far, when we update the road segment with an updated speed, it doesnt take into account the direction. For example, A to B, and B to A is being updated with the same speed which may not always the case. Is there a way to dig deeper to update the road segments with different speeds?

**Reply by nilsnolde** [reply]
You need to update both edges in both directions. Each bidirectional road segments has 2 graph edges, seems you’re only updating one so far. We can’t outline the process for you, as we don’t know your data. Feel free to drop me a mail on enquiry@gis-ops.com for commercial support.

**Reply by ChouBokYann** [reply]
Thank you for the fast response! Referring to the "way_to_edges" file found in https://github.com/valhalla/valhalla/blob/master/src/mjolnir/valhalla_ways_to_edges.cc, does that mean that once ive inputted my way_id into this file. I must also find the proper directed edge and then input the speed then?

**Reply by nilsnolde** [reply]
Yeah, you’ll need to have one entry for each directed edge eventually in the traffic CSV tiles.

**Reply by ChouBokYann** [reply]
I saw on issue #2598 that these edge IDs are unstable and will change overtime. Does that mean that everytime I want to update the speeds, I will need to first query all the new edge IDs? Also, is there already a way in Valhalla that allows us to split the edge IDs by their direction. From my understanding, once ive inputted the way_id into the ways_to_edges file, it returns all the edge_ids, disregarding their direction. Is there currently a way to split them up by their direction? (i.e. side of street, heading)

So sorry for the spam of questions! and Thank you so much for your time and patience.

**Reply by nilsnolde** [reply]
> I saw on issue https://github.com/valhalla/valhalla/issues/2598 that these edge IDs are unstable and will change overtime.

They don't change over time, they only change on new graph builds. 

> Does that mean that everytime I want to update the speeds, I will need to first query all the new edge IDs?

Currently you'll have to build a new graph actually. We're not robust to "update" speeds for historical traffic without rebuilding the graph (IRL with commercial datasets this doesn't make sense, but could of course for more dynamic datasets). I think it wouldn't be hard to add support for that, i.e. just run `valhalla_add_predicted_traffic` on an already "loaded" graph, where it removes old traffic data, resets all speeds on the directed edges, then just adds the new batch. But this isn't possible today.

For very frequent speed updates, our live traffic infrastructure is more suitable. That's a much bigger lift though.

> it returns all the edge_ids, disregarding their direction

? a directed edge has a direction. it's your job to use this output and correlate it to your speed data (which needs to have speeds according to each direction as well of course). How to do that is beyond the scope of free support IMO as it highly depends on your actual data/workflow.

**Reply by ChouBokYann** [reply]
If I were to collaborate and implement the function, how would I go about doing it? What are the class methods that I should modify?

Thanks in advance : )
